File: easy/0604_design_compressed_string_iterator/solution.cpp
using namespace std;

#include <string>
#include <vector>

class StringIterator {
private:
  vector<pair<char, int>> encodedRuns;

public:
  StringIterator(string compressedString) {
    int prev = compressedString.size();
    int curr = prev - 1;

    while (curr >= 0) {
      while (isdigit(compressedString[curr]))
        --curr;

      encodedRuns.push_back(
          {compressedString[curr],
           stoi(compressedString.substr(curr + 1, prev - curr - 1))});

      prev = curr--;
    }
  }

  char next() {
    if (encodedRuns.size() == 0)
      return ' ';

    --encodedRuns.back().second;
    auto &[ch, count] = encodedRuns.back();

    if (count == 0)
      encodedRuns.pop_back();

    return ch;
  }

  bool hasNext() { return encodedRuns.size() > 0; }
};

File: easy/0604_design_compressed_string_iterator/solution.py
class StringIterator:
    def __init__(self, compressed_string: str):
        self.__encoded_runs = []

        prev = len(compressed_string)
        curr = prev - 1

        while curr >= 0:
            while compressed_string[curr].isdigit():
                curr -= 1

            self.__encoded_runs.append(
                [
                    compressed_string[curr],
                    int(compressed_string[curr + 1 : prev]),
                ]
            )

            prev = curr
            curr -= 1

    def next(self) -> str:
        if not self.__encoded_runs:
            return " "

        self.__encoded_runs[-1][1] -= 1
        c, count = self.__encoded_runs[-1]

        if count == 0:
            self.__encoded_runs.pop()

        return c

    def hasNext(self) -> bool:
        return bool(self.__encoded_runs)
